tools/congestion-model/src/strategy/traffic_light.rs

```diff
@@ -0,0 +1,228 @@
+            // shard's color note that this is "attached gas" and the actual gas
+            // the receiver will need to burn while executing can be much lower
+            green_outgoing_receipts_gas: 30_000 * TGAS, // ~30 chunks
+            amber_outgoing_receipts_gas: 5_000 * TGAS,  // ~5 chunks
+            // RED: only the selected shard can send this amount. This number
+            // should be quite low. The receiving shard is not running out of
+            // executable receipts. But it should also not be too slow, since
+            // this is all the progress we get on the sending shard in case it
+            // is also itself full .
+            red_outgoing_receipts_gas_single_shard: 1_000 * TGAS, // < 1 chunk
+
+            // how much gas is assigned to new transactions on a shard depending on it color
+            green_new_tx_gas: TX_GAS_LIMIT,
+            amber_new_tx_gas: 100 * TGAS,
+            red_new_tx_gas: 20 * TGAS,
+
+            // state to be initialized in `init`
+            shard_id: None,
+            outgoing_buffer: HashMap::new(),
+            all_shards: vec![],
+
+            // state to update every round
+            outgoing_gas_allowance: HashMap::new(),
+        }
+    }
+}
+
+impl CongestionStrategy for TrafficLight {
+    fn init(
+        &mut self,
+        own_id: crate::ShardId,
+        all_shards: &[crate::ShardId],
+        queue_factory: &mut dyn QueueFactory,
+    ) {
+        self.shard_id = Some(own_id);
+        self.outgoing_buffer = all_shards
+            .iter()
+            .map(|&id| (id, queue_factory.register_queue(own_id, &format!("out_buffer_{id}"))))
+            .collect();
+        self.all_shards = all_shards.to_vec();
+    }
+
+    fn compute_chunk(&mut self, ctx: &mut ChunkExecutionContext) {
+        let tx_gas = match self.my_prev_status(ctx) {
+            TrafficLightStatus::Green => self.green_new_tx_gas,
+            TrafficLightStatus::Amber => self.amber_new_tx_gas,
+            TrafficLightStatus::Red { .. } => self.red_new_tx_gas,
+        };
+        self.reset_outgoing_gas_allowance(ctx);
+
+        // try forwarding buffered outgoing receipts
+        for (shard, &q) in &self.outgoing_buffer {
+            while !(&mut ctx.queue(q)).is_empty()
+                && self.outgoing_gas_allowance[&shard]
+                    >= (&mut ctx.queue(q)).front().unwrap().attached_gas
+            {
+                let receipt = (&mut ctx.queue(q)).pop_front().unwrap();
+                let allowance = self.outgoing_gas_allowance.get_mut(&shard).unwrap();
+                *allowance = allowance.saturating_sub(receipt.attached_gas);
+                ctx.forward_receipt(receipt);
+            }
+        }
+
+        while ctx.gas_burnt() < tx_gas {
+            if let Some(tx) = ctx.incoming_transactions().pop_front() {
+                let receiver = ctx.tx_receiver(tx);
+                if status(ctx, receiver) != TrafficLightStatus::Green {
```

Comment:
just sharing thoughts no suggestions here
This is an excellent heuristic for the typical traffic but a terrible one when dealing with malicious attacks. We should make sure that we don't get a false sense of security based on how well this works on simple transactions. It's great that you've added all the multi hop loads that should bypass this optimization easily.